Make getting and setting the RAND default method thread safe

levitte opened the pull request:
From a comment in #3136, where it was noticed that `RAND_set_rand_method()` and friends weren't thread safe.

levitte committed:
Make getting and setting the RAND default method thread safe

mattcaswell commented on crypto/rand/rand_lib.c:
```diff
@@ -12,27 +12,49 @@
+
+DEFINE_RUN_ONCE_STATIC(do_rand_lock_init)
+{
+    OPENSSL_init_crypto(0, NULL);
```
Why is this necessary? It looks like a potential source of deadlocks

levitte replied:
Perhaps it isn't...  I copied this stoopidly from the one in `md_rand.c`...

mattcaswell commented on crypto/rand/rand_lib.c:
```diff
@@ -49,13 +71,18 @@ const RAND_METHOD *RAND_get_rand_method(void)
             default_RAND_meth = RAND_OpenSSL();
     }
+    CRYPTO_THREAD_unlock(rand_meth_lock);
     return default_RAND_meth;
```
Are assignments atomic? This read of the variable is unprotected by a lock, so I wonder what happens if RAND_set_rand_method() is executed at the same time as this read is taking place.

levitte replied:
It's protected by the read on line 58, no?

levitte replied:
Or oh, RAND_*set*_rand_method()!  Ok, I get what you mean

mattcaswell reviewed (changes requested):
I wonder what the performance impact of this is?

levitte commented:
Fixed all outstanding concerns

> I wonder what the performance impact of this is?

Good question

mattcaswell commented:
Travis is showing a failure.

mattcaswell commented on crypto/rand/rand_lib.c:
```diff
@@ -49,13 +72,19 @@ const RAND_METHOD *RAND_get_rand_method(void)
             default_RAND_meth = RAND_OpenSSL();
     }
-    return default_RAND_meth;
+    tmp_meth = default_RAND_meth;
```
Does default_RAND_meth have to be declared volatile for this to work? Can the compiler "optimise" this solution away? I'm not really sure.

kaduk replied:
I believe that the function call to unlock will provide a compiler-level barrier (and the lock itself will give a memory barrier at runtime).  Function calls are considered to have side effects, in the abstract machine.

levitte committed:
[to be squashed] remember to free the locks on cleanup

levitte commented:
> Travis is showing a failure.

Yeah, I forgot to free the locks...

levitte commented:
You know, this isn't done anywhere else (have a look at `crypto/rsa/rsa_lib.c` for example), and back in #3136, @richsalz makes a good point reminding us of #2244.  Closing this.

mattcaswell commented:
I'm confused. The discussion in #2244 seems to suggest that this PR has it exactly right and is what we should be doing??

levitte commented:
Maybe I'm confused too, I thought it about redoing the whole defaults stuff to happen mostly at compile time...

levitte committed:
Don't try to clean up RAND from ENGINE

This is especially harmful since OPENSSL_cleanup() has already called
the RAND cleanup function

mattcaswell commented:
Travis is now passing. LGTM. My only outstanding question is my one on volatile above.

kaduk reviewed (commented):
> redoing the whole defaults stuff to happen mostly at compile time...

There's using compile-time initialization instead of CRYPTO_THREAD_run_once(), and then there's having the default_foo variable for what is the default method when a method is not explicitly specified.  The latter makes sense to allow changing at runtime.

levitte commented:
So, errr, was that an approval hidden in there somewhere?

kaduk commented on crypto/rand/rand_lib.c:
```diff
@@ -12,27 +12,50 @@
 static ENGINE *funct_ref = NULL;
+static CRYPTO_RWLOCK *rand_engine_lock = NULL;
 #endif
 static const RAND_METHOD *default_RAND_meth = NULL;
```
In other projects I usually see a comment "requires rand_meth_lock for access" or semantically similar; do we want to do something like that?

kaduk reviewed (approved):
Sure, you can count it as an approval from me if you want.  (Not that it adds up to anything concrete, of course.)

levitte commented:
Merged.

master: 87975cfa91c3f06a52f2550ed2ef632644be6dde and 789a2b6250d5e05dfde6ce259e79ef8c172c9f3f
1.1.0: 67114d93b47aa02192272182595c2b5d1bba8298 and df13eb7ae70073de8c95ca3c7d2f587037582470
